### PR TITLE
WIP: add pytest-httpbin

### PR DESCRIPTION
**Pull request by kevin1024**
This adds support to most of the tests that rely on httpbin.  There are still a few that need to hit the internet.  Not sure if I can make the entire test suite run offline or not.  Might need to mock DNS for this to work.

### CODE REVIEW

**Comment by IvanMalison** [comment]
This is awesome. How much more work needs to be done here?

**Comment by kevin1024** [comment]
Well, I was hoping to make the test suite entirely offline but if I give up on that goal, this probably just needs a rebase and we can at least get it to 90% offline.  I will see what I can do this weekend.

**Comment by IvanMalison** [comment]
whats hard about the last 10%?

File: README.md [hunk 1]
```diff
@@ -420,6 +420,10 @@ documentation](http://boto.readthedocs.org/en/latest/getting_started.html) for
 how to set this up. I have marked the boto tests as optional in Travis so you
 don't have to worry about them failing if you submit a pull request.
 
+Since VCR.py  uses [pytest-httpbin](https://github.com/kevin1024/pytest-httpbin) for its test suite, you need to set an
```
Comment (jayvdb) [inline README.md]: two consecutive spaces, and extra long line should be wrapped ~80cpl

File: tests/integration/test_config.py [hunk 2]
```diff
@@ -5,46 +5,46 @@
...
+def test_override_match_on(tmpdir, httpbin):
     my_vcr = vcr.VCR(match_on=['method'])
 
     with my_vcr.use_cassette(str(tmpdir.join('test.json'))):
-        urlopen('http://httpbin.org/')
+        urlopen(httpbin.url + '/')
 
     with my_vcr.use_cassette(str(tmpdir.join('test.json'))) as cass:
-        urlopen('http://httpbin.org/get')
+        urlopen(httpbin.url + '/')
```
Comment (jayvdb) [inline tests/integration/test_config.py]: /get ?

File: tests/integration/test_ignore.py [hunk 3]
```diff
@@ -2,54 +2,73 @@
...
+        if args[0] in overrides:
+            address = overrides[args[0]]
+            return [(2, 1, 6, '', (address, args[1]))]
+        return real_getaddrinfo(*args, **kwargs)
+    socket.getaddrinfo = fake_getaddrinfo
+    yield
+    socket.getaddrinfo = real_getaddrinfo
+
+def test_ignore_localhost(tmpdir, httpbin):
+    with overridden_dns({'httpbin.org':'127.0.0.1'}):
```
Comment (jayvdb) [inline tests/integration/test_ignore.py]: {'httpbin.org':'127.0.0.1'} -> {'httpbin.org': '127.0.0.1'}
(here and below)
probably other pep8 issues also

File: tests/integration/test_ignore.py [hunk 4]
```diff
@@ -2,54 +2,73 @@
...
+
+@contextmanager
+def overridden_dns(overrides):
+    """
+    Monkeypatch socket.getaddrinfo() to override DNS lookups (name will resolve
+    to address)
+    """
+    real_getaddrinfo = socket.getaddrinfo
+    def fake_getaddrinfo(*args, **kwargs):
+        if args[0] in overrides:
```
Comment (jayvdb) [inline tests/integration/test_ignore.py]: would aid understanding if you added
`foo, bar = args[0], args[1]`
(or expanded args in the arguments list)

File: tests/integration/test_requests.py [hunk 5]
```diff
@@ -197,12 +191,12 @@ def test_nested_cassettes_with_session_created_before_nesting(scheme, tmpdir):
         assert session.get(url).content == third_body
 
     # Make sure that the session can now get content normally.
-    session.get('http://www.reddit.com')
+    session.get(httpbin_both.url)
```
Comment (jayvdb) [inline tests/integration/test_requests.py]: maybe..?  to closely match the prior intention, request something else, like `httpbin_both + '/robots.txt'` and assert it contains `User-agent: *`

File: tests/integration/test_urllib2.py [hunk 6]
```diff
@@ -4,125 +4,127 @@
...
 from six.moves.urllib.request import urlopen
 from six.moves.urllib_parse import urlencode
+import pytest_httpbin.certs
 
 # Internal imports
 import vcr
 
 from assertions import assert_cassette_has_one_response
 
+def open(*args, **kwargs):
```
Comment (jayvdb) [inline tests/integration/test_urllib2.py]: `open` -> `urlopen_with_cafile` ?

File: tests/integration/test_wild.py [hunk 7]
```diff
@@ -40,11 +40,13 @@ def _pretend_to_be_flickr_library():
         data = r.read()
         h.close()
 
-    with vcr.use_cassette('fixtures/vcr_cassettes/flickr.yaml') as cass:
+    c_file = str(tmpdir.join('fixtures/vcr_cassettes/flicr.yaml'))
```
Comment (jayvdb) [inline tests/integration/test_wild.py]: `flicr.yaml` -> `flickr.yaml`

and it looks like this is a separate bugfix, which could be submitted separately.

**Comment by kevin1024** [comment]
@jayvdb thanks so much for the review!

File: .travis.yml [hunk 8]
```diff
@@ -38,7 +38,7 @@ python:
 - 3.4
 - pypy
 install:
-- pip install tox-travis
+- pip install tox-travis --use-mirrors
```
Comment (jayvdb) [inline .travis.yml]: This option is deprecated, and removed from the version of pup that is installed in the '3.5' image.

**Comment by jayvdb** [comment]
There are lots of lines using `httpbin.url + '/'`.

Is the `+ '/'` necessary?  I'm guessing it makes the URL precisely the same as the previous implementation, but it should be functionally equivalent without the `+ '/'`, and a little neater.

File: README.md [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+If you made a custom serializer, you will need to update it to match the new
+API in version 1.0.x
+
+ * Serializers now take dicts and return strings.
+ * Deserializers take strings and return dicts (instead of requests, responses
+   pair)
+
+
+## Changelog
+  * 1.4.1 Include README, tests, LICENSE in package.  Thanks @ralphbean.
```
Comment (jayvdb) [inline README.md]: color me confused.
docs/changelog.rst goes up to 1.7.4

File: README.md [hunk 10]
```diff
@@ -0,0 +1,599 @@
+# VCR.py
```
Comment (jayvdb) [inline README.md]: This new `README.md` appears to be large chunks of `docs/` pulled together.  Sifting through it, I can only see one small note regarding `pytest.httpbin`, which probably should also have an equivalent change to `docs/`.
Anyway, im a bit lost about this new file.

Comment (kevin1024) [inline README.md]: Yeah it was a merge conflict screwup.  Good catch.

**Comment by jayvdb** [comment]
> WIP: add pytest-httpbin

still a "WIP" ? ;-)

Anyway, thx so much.  This should make development much easier.

It is interesting that the travis build time was larger than previous `master` builds.  Maybe due to temporary issues with Travis CI; if not temporary, maybe we can find some ways to speed it up.

**Comment by jayvdb** [comment]
Oh, addition of pypy3 and py3.5 are to blame for the increased build times.  silly me.

**Comment by kevin1024** [comment]
Thanks for the review and the encouragement :+1: Glad to get this merged.

**Pull request merged**